src/Testable/Navigation.elm

```diff
@@ -0,0 +1,156 @@
+module Testable.Navigation exposing (..)
+
+import List.Extra
+import
+    -- This "unused" import is required for the native bindings
+    Native.Testable.Navigation
+import Navigation exposing (Location)
+import Regex exposing (HowMany(..), find, regex)
+
+
+type alias History =
+    ( Int, List Location )
+
+
+type Msg
+    = Jump Int
+    | New String
+    | Modify String
+    | Visit String
+
+
+type ReturnMsg
+    = NoOp
+    | ReturnLocation Location
+    | TriggerLocationMsg Location
+
+
+currentLocation : History -> Location
+currentLocation ( index, history ) =
+    List.Extra.getAt index history
+        |> Maybe.withDefault (getLocation "")
+
+
+update : Msg -> History -> ( History, ReturnMsg )
+update msg ( index, history ) =
+    let
+        location =
+            currentLocation ( index, history )
+    in
+    case msg of
+        Jump n ->
+            let
+                nextIndex =
+                    min (max (index + n) 0) (List.length history - 1)
+
+                nextLocation =
+                    List.Extra.getAt nextIndex history
+                        |> Maybe.withDefault (getLocation "")
+            in
+            ( ( nextIndex, history ), TriggerLocationMsg nextLocation )
+
+        New url ->
+            let
+                nextLocation =
+                    setLocation url location
+
+                modifiedHistory =
+                    List.take (index + 1) history ++ [ nextLocation ]
+            in
+            ( ( index + 1, modifiedHistory ), ReturnLocation nextLocation )
+
+        Modify url ->
+            let
+                nextLocation =
+                    setLocation url location
+
+                modifiedHistory =
+                    history
+                        |> List.Extra.updateIfIndex ((==) index) (always nextLocation)
+
+                effect =
+                    if nextLocation == location then
+                        NoOp
+                    else
+                        ReturnLocation nextLocation
+            in
+            ( ( index, modifiedHistory ), effect )
+
+        Visit url ->
+            let
+                ( modifiedHistory, _ ) =
+                    update (New url) ( index, history )
+            in
+            ( modifiedHistory, TriggerLocationMsg (currentLocation modifiedHistory) )
+
+
+init : History
+init =
+    ( 0, [ initialLocation ] )
+
+
+initialLocation : Location
+initialLocation =
+    getLocation "https://elm.testable/"
```

Comment:
Is there a nice way we could make the API require the test author to provide the initial location iff the program being tested is a `Navigation.program` and not have a default provided?

Reply:
Ok, I've added `startWithLocation` and `startWithFlagsAndLocation`. This increased complexity a bit, because now we need to save the original `init` function, that's because when the user calls `Navigation.Program` (which is before using `TestContext.start`) the init function is called with the current url.

So I store the original init function and call it again with the url from `startWithLocation`.

Still, it is possible for the user to use the `Navigation.Program` with the `TestContext.start` function (without the withLocation) because like I said, the types signatures are the same. In this case, the url fallbacks to `https://elm.testable/`

What do you think?